qa/evil-tests/src/test/java/org/elasticsearch/bootstrap/EvilSpawnerTests.java

```diff
@@ -0,0 +1,164 @@
+        + "\n"
+        + "int main(int, char **) {\n"
+        + "    const char *tmpDir = ::getenv(\"TMPDIR\");\n"
+        + "    if (tmpDir == 0) {\n"
+        + "        return EXIT_FAILURE;\n"
+        + "    }\n"
+        + "\n"
+        + "    std::string helloFile(tmpDir);\n"
+        + "    helloFile += '/';\n"
+        + "    helloFile += \"" + HELLO_FILE_NAME + "\";\n"
+        + "    std::string goodbyeFile(tmpDir);\n"
+        + "    goodbyeFile += '/';\n"
+        + "    goodbyeFile += \"" + GOODBYE_FILE_NAME + "\";\n"
+        + "\n"
+        + "    std::ofstream helloStrm(helloFile.c_str());\n"
+        + "    helloStrm << \"Hello, world!\\n\";\n"
+        + "    helloStrm.close();\n"
+        + "\n"
+        + "    // Assuming no input on stdin, this will pause the program until end-of-file on stdin\n"
+        + "    char c;\n"
+        + "    std::cin >> c;\n"
+        + "\n"
+        + "    std::ofstream goodbyeStrm(goodbyeFile.c_str());\n"
+        + "    goodbyeStrm << \"Goodbye, world!\\n\";\n"
+        + "    goodbyeStrm.close();\n"
+        + "\n"
+        + "    return EXIT_SUCCESS;\n"
+        + "}\n";
+
+    public void testControllerSpawn() throws IOException, InterruptedException {
+        Path esHome = createTempDir().resolve("esHome");
+        Settings.Builder settingsBuilder = Settings.builder();
+        settingsBuilder.put(Environment.PATH_HOME_SETTING.getKey(), esHome.toString());
+        Settings settings = settingsBuilder.build();
+
+        Environment environment = new Environment(settings);
+
+        // This plugin will have a controller daemon built for it
+        Path plugin = environment.pluginsFile().resolve("test_plugin");
+        Files.createDirectories(plugin);
+        Path controllerProgram = Spawner.makeSpawnPath(plugin);
+        if (compileControllerProgram(environment, controllerProgram) == false) {
+            // Don't fail the test if there's an error compiling the plugin - the build machine probably doesn't have g++ installed
+            Loggers.getLogger(EvilSpawnerTests.class).warn("Could not compile native controller program for testing");
+            return;
+        }
+        Loggers.getLogger(EvilSpawnerTests.class).info("Successfully compiled native controller program for testing");
+
+        // This plugin will NOT have a controller daemon
+        Path otherPlugin = environment.pluginsFile().resolve("other_plugin");
+        Files.createDirectories(otherPlugin);
+
+        Path helloFile = environment.tmpFile().resolve(HELLO_FILE_NAME);
+        Path goodbyeFile = environment.tmpFile().resolve(GOODBYE_FILE_NAME);
+
+        // Clean up any files left behind by previous debugging
+        Files.deleteIfExists(helloFile);
+        Files.deleteIfExists(goodbyeFile);
+
+        Spawner spawner = new Spawner();
+        spawner.spawnNativePluginControllers(environment);
+
+        try {
+            // Give the program time to start up
+            Thread.sleep(500);
```

Comment:
instead of adding a sleep here you can do this

```Java
assertBusy(() -> {
         List<OutputStream> stdinReferences = spawner.getStdinReferences();
           // 1 because there should only be a reference in the list for the plugin that had the controller daemon, not the other plugin
          assertEquals(1, stdinReferences.size());
          assertTrue(Files.isRegularFile(helloFile));
          assertFalse(Files.exists(goodbyeFile));
          List<String> lines = Files.readAllLines(helloFile, StandardCharsets.UTF_8);
          assertFalse(lines.isEmpty());
          assertEquals("Hello, world!", lines.get(0));
});
spawner.close();
assertBusy(() -> {
         // Give the program time to exit
         assertTrue(Files.isRegularFile(goodbyeFile));
         lines = Files.readAllLines(goodbyeFile, StandardCharsets.UTF_8);
         assertFalse(lines.isEmpty());
         assertEquals("Goodbye, world!", lines.get(0));
});
```

`assertBusy` will try to execute the code and pause for an quadratically growing amount of time between retries. if it fails after retrying N times it will just rethrow the exception it's much more robust and will also pass on slowish machines